[Qt] replace fee slider with a Dropdown, extend conf. targets

jonasschnelli opened the pull request:
Addresses #10590.

This PR replaces the fee slider with a Dropbox box. The Dropdown contains the target in ~blocks and ~estimated time to confirm.

The current supported confirmation targets are: 2, 4, 6, 12, 48, 144, 504, 1008

<img width="865" alt="bildschirmfoto 2017-07-07 um 21 56 19" src="https://user-images.githubusercontent.com/178464/27974951-83e9f284-6360-11e7-8c45-d15e3be88add.png">

jonasschnelli committed:
[Qt] replace fee slider with a Dropdown, extend conf. targets

jonasschnelli commented:
I probably need to mention why a dropbox and not a slider.

* Sliders tend to be linear (otherwise nobody would understand them and I bet it would require a custom, non standard UI object) which would make it useless if we would support conf. targets up to 1008.
* A dropbox allows one to give a better overview (slider: you only see the underlaying value if you change it).
* Dropbox would allow to show the feerate and or a text ("fast", "economy", etc,), ... and eventual the absolute fee by running the coin-selections if the parameters are provided.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -31,6 +31,25 @@
 #include <QTextDocument>
 #include <QTimer>
 
+static const std::array<int, 8> confTargets = { {2, 4, 6, 12, 48, 144, 504, 1008} };
```
Rename `conf_targets`.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -31,6 +31,25 @@
 
+static const std::array<int, 8> confTargets = { {2, 4, 6, 12, 48, 144, 504, 1008} };
+int getConfTargetForIndex(int index) {
+    if (index+1 > static_cast<int>(confTargets.size())) {
```
```cpp
if (index >= ...
```

promag replied:
Or just:
```cpp
  return conf_targets[max(0, min(conf_targets.size() - 1, index)]
```

flack commented:
Looking at the screenshot, I would drop the tilde before the number of blocks, i.e. `2 Blocks, ~20 minutes` instead of `~2 Blocks, ~20 minutes`. It already says "confirmation target", so it should be obvious that it's not exactly x blocks. In other words, make it say "target x blocks (approx. y minutes)" instead of "target approx. x blocks (approx. y minutes)".

Also, shouldn't blocks have a lowercase b?

morcos commented:
Will review shortly.  Thanks for doing!

Also please add a target for 24 for compatibility with the old estimates (where 25 was the highest possible target).  It may be a target that people are used to using.

luke-jr commented:
Confirmation is when it's 6 blocks deep, so +5 to all the numbers here? Or change it to "begin confirmation" like we currently have...

gmaxwell commented:
I like how it looks, @morcos any concerns about missing 3,5 making users fee choices more clumpy?

jonasschnelli commented:
Re-added targets 3, 5, 24 and removed the tilde in front of the target-in-blocks/-time.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -177,10 +199,10 @@ void SendCoinsDialog::setModel(WalletModel *_model)
         // set the smartfee-sliders default value (wallets default conf.target or last stored value)
         QSettings settings;
-        if (settings.value("nSmartFeeSliderPosition").toInt() == 0)
-            ui->sliderSmartFee->setValue(ui->sliderSmartFee->maximum() - model->getDefaultConfirmTarget() + 2);
```
Should call `settings.remove("nSmartFeeSliderPosition")`?

jonasschnelli replied:
Hmm... I think an addition is required: we probably need to migrate the old slider value to the new dropdown?

morcos replied:
That doesn't seem particularly important to me.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -152,9 +171,12 @@ void SendCoinsDialog::setModel(WalletModel *_model)
-        connect(ui->sliderSmartFee, SIGNAL(valueChanged(int)), this, SLOT(updateGlobalFeeVariables()));
-        connect(ui->sliderSmartFee, SIGNAL(valueChanged(int)), this, SLOT(coinControlUpdateLabels()));
+        for (const int &n : confTargets) {
+            ui->confTargetSelector->addItem(tr("%1 blocks, %2").arg(n).arg(GUIUtil::formatNiceTimeOffset(n*Params().GetConsensus().nPowTargetSpacing)));
```
Alternative format: `"%2 (%1 blocks)"` since the label is `Confirmation time`.

jonasschnelli replied:
Yes. Your right.. will fix.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -31,6 +31,25 @@
 #include <QTimer>
 
+static const std::array<int, 11> confTargets = { {2, 3, 4, 5, 6, 12, 24, 48, 144, 504, 1008} };
+int getConfTargetForIndex(int index) {
```
Update block style. Same for `getIndexForConfTarget`.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -31,6 +31,25 @@
 #include <QTextDocument>
 #include <QTimer>
 
+static const std::array<int, 11> confTargets = { {2, 3, 4, 5, 6, 12, 24, 48, 144, 504, 1008} };
```
Snake case.

TheBlueMatt replied:
nit: not sure its worth having 5 here, maybe not 3 either.

TheBlueMatt replied:
Also, why the extra set of {}?

jonasschnelli replied:
> Also, why the extra set of {}?

Compiler warning.

morcos commented:
ACK 19eae9c

@gmaxwell Actually my preference would be to remove the 3 and 5 targets.  I think it gives a false sense of accuracy with the estimates.  I don't imagine much clumping concern as those estimates often aren't that different from each other and there is a much larger range of estimates to choose from as well as time variability.  I also think its a bit cleaner not to be faced with too many choices in a dropdown.

TheBlueMatt commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -31,6 +31,25 @@
+            return i;
+        }
+    }
+    return confTargets.back();
```
I think you meant .size() - 1.

jonasschnelli replied:
Oh. Yes. Will fix.

TheBlueMatt reviewed (commented):
Really think we need this for 15.

sipa commented:
GUI-screenshot-looks-nice-ACK. Didn't look at the code.

achow101 commented:
utACK 19eae9c3602c784d55f8edc94300da1ece686a6c modulo the thing @TheBlueMatt pointed out.

I don't think we need the 3, 4, or 5 block target estimates.

It may also be useful to include a tooltip that explains why you see the line "Estimated to begin confirmation within N blocks" where N is less than the confirmation target that you chose as seen in the screenshot below. This discrepancy may confuse users.

![image](https://user-images.githubusercontent.com/3782274/28143521-d5348182-671a-11e7-9476-f4ab3436098a.png)

morcos commented:
>  tooltip that explains why you see the line "Estimated to begin confirmation within N blocks" where N is less than the confirmation target that you chose

Yes would be a nice improvement, could be left for another PR as it's a preexisting issue.  Would be great to get than in for 0.15 as well, but wouldn't want it to hold up this.

The other feature that I'd ideally like is the ability to force estimates conservative or economical regardless of the Replace-By-Fee setting.  Again just a wish list item.

jonasschnelli committed:
[Qt] migrate old fee slider value to new dropbown
Always round up (conservative)

jonasschnelli commented:
Removed again target 3 and 5 and fixed points found by @TheBlueMatt and @promag.
Added another commit that add a tiny migration logic.

Thanks for a retest/re-ack.

morcos commented:
Tested ACK 2aef1f1

TheBlueMatt commented:
utACK 2aef1f18296fcd3aa3c91afdf152add8a8e80bd4
